File: 0001-0500/0226-Invert-Binary-Tree/java-0226/src/Solution2.java
/// Source : https://leetcode.com/problems/invert-binary-tree/description/
/// Author : liuyubobobo
/// Time   : 2017-11-17

import java.util.LinkedList;

/// Non-Recursive
/// Time Complexity: O(n), where n is the node's number of the tree
/// Space Complexity: O(h), where h is the height of the tree
public class Solution2 {

    // Definition for a binary tree node.
    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode(int x) { val = x; }
    }

    public TreeNode invertTree(TreeNode root) {

        if(root == null)
            return null;

        LinkedList<TreeNode> queue = new LinkedList<TreeNode>();
        queue.addLast(root);
        while(!queue.isEmpty()){
            TreeNode curNode = queue.removeFirst();

            TreeNode tempNode = curNode.left;
            curNode.left = curNode.right;
            curNode.right = tempNode;

            if(curNode.left != null)
                queue.addLast(curNode.left);
            if(curNode.right != null)
                queue.push(curNode.right);
        }

        return root;
    }
}

File: 0001-0500/0226-Invert-Binary-Tree/py-0226/Solution2.py
# Source : https://leetcode.com/problems/invert-binary-tree/
# Author : penpenps
# Time   : 2019-08-01

# BFS, iterative 
# Time Complexity: O(n), the number of tree's nodes
# Space Complexity: O(n), the max node number by level

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def invertTree(self, root: TreeNode) -> TreeNode:
        if not root: return root
        queue = [root]
        while queue:
            node = queue.pop(0)
            node.left, node.right = node.right, node.left
            if node.left:
                queue.append(node.left)
            if node.right:
                queue.append(node.right)
        return root
    